pkg/lib/cockpit/dialog.tsx

```diff
@@ -0,0 +1,1554 @@
+            if (k.indexOf(from_path + ".") == 0) {
+                const to = to_path + k.substring(from_path.length);
+                debug("rename", k, to);
+                this.#validation_state[to] = this.#validation_state[k];
+                this.#validation_state[to].path = to;
+                delete this.#validation_state[k];
+            }
+        }
+    }
+
+    /* The first thing run_action does is to trigger a new validation
+       round and then wait for all the asynchronous results to have
+       come in.
+
+       If there are any DialogValidationState objects that are waiting
+       for a timeout, we want to cancel those and start over, so that
+       their validation starts immediately. (Also, it would be hairy
+       to wait for those timeouts to be over from here.)
+     */
+
+    async validate(): Promise<boolean> {
+        this.#cancel_all_validation_timeouts();
+        this.#trigger_validation();
+        await this.#wait_for_validation_promises();
+        return !this.#validation_failed;
+    }
+
+    #cancel_all_validation_timeouts() {
+        for (const p in this.#validation_state) {
+            const state = this.#validation_state[p];
+            if (state.timeout_id) {
+                debug("timeout bulk cancel", p);
+                window.clearTimeout(state.timeout_id);
+                delete this.#validation_state[p];
+            }
+        }
+    }
+
+    async #wait_for_validation_promises(): Promise<void> {
+        for (const path in this.#validation_state) {
+            const state = this.#validation_state[path];
+            if (state.promise) {
+                debug("waiting for promise", path);
+                await state.promise;
+                debug("waiting for promise done", path);
+            }
+        }
+    }
+
+    async run_action(func: (vals: V) => Promise<void>) {
+        this.error = null;
+        this.#action_running = true;
+        this.#update();
+        if (!await this.validate()) {
+            this.#action_running = false;
+            this.#update();
+            return;
+        }
+
+        try {
+            await func(this.values);
+        } catch (ex) {
+            console.error(String(ex));
+            this.error = ex;
+        }
+
+        this.#action_running = false;
+        this.#update();
+    }
+
+    top_value(update_func?: ((val: V) => void) | undefined): DialogValue<V> {
+        return new DialogValue<V>(
+            this as DialogState<unknown>,
+            () => this.values,
+            (val) => {
+                debug("set", val);
+                if (this.#action_running) {
+                    debug("set denied");
+                    return;
+                }
+                this.values = val;
+                this.#update();
+                if (this.#online_validation)
+                    this.#trigger_validation();
+                if (update_func)
+                    update_func(val);
```

Comment:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22700-8ebf59dd-20251219-120533-fedora-43-devel-other/Coverage/lcov/github-pr.diff.gcov.html)